test_libs/pyspec/tests/test_finality.py

```diff
@@ -0,0 +1,193 @@
+from copy import deepcopy
+
+import pytest
+
+import eth2spec.phase0.spec as spec
+
+from eth2spec.phase0.state_transition import (
+    state_transition,
+)
+from .helpers import (
+    build_empty_block_for_next_slot,
+    fill_aggregate_attestation,
+    get_current_epoch,
+    get_epoch_start_slot,
+    get_valid_attestation,
+    next_epoch,
+)
+
+# mark entire file as 'state'
+pytestmark = pytest.mark.state
+
+
+def check_finality(state,
+                   prev_state,
+                   current_justified_changed,
+                   previous_justified_changed,
+                   finalized_changed):
+    if current_justified_changed:
+        assert state.current_justified_epoch > prev_state.current_justified_epoch
+        assert state.current_justified_root != prev_state.current_justified_root
+    else:
+        assert state.current_justified_epoch == prev_state.current_justified_epoch
+        assert state.current_justified_root == prev_state.current_justified_root
+
+    if previous_justified_changed:
+        assert state.previous_justified_epoch > prev_state.previous_justified_epoch
+        assert state.previous_justified_root != prev_state.previous_justified_root
+    else:
+        assert state.previous_justified_epoch == prev_state.previous_justified_epoch
+        assert state.previous_justified_root == prev_state.previous_justified_root
+
+    if finalized_changed:
+        assert state.finalized_epoch > prev_state.finalized_epoch
+        assert state.finalized_root != prev_state.finalized_root
+    else:
+        assert state.finalized_epoch == prev_state.finalized_epoch
+        assert state.finalized_root == prev_state.finalized_root
+
+
+def next_epoch_with_attestations(state,
+                                 fill_cur_epoch,
+                                 fill_prev_epoch):
+    post_state = deepcopy(state)
+    blocks = []
+    for slot in range(spec.SLOTS_PER_EPOCH):
+        block = build_empty_block_for_next_slot(post_state)
+        if fill_cur_epoch:
+            slot_to_attest = post_state.slot - spec.MIN_ATTESTATION_INCLUSION_DELAY + 1
+            if slot_to_attest >= get_epoch_start_slot(get_current_epoch(post_state)):
+                cur_attestation = get_valid_attestation(post_state, slot_to_attest)
+                fill_aggregate_attestation(post_state, cur_attestation)
+                block.body.attestations.append(cur_attestation)
+
+        if fill_prev_epoch:
+            slot_to_attest = post_state.slot - spec.SLOTS_PER_EPOCH + 1
+            prev_attestation = get_valid_attestation(post_state, slot_to_attest)
+            fill_aggregate_attestation(post_state, prev_attestation)
+            block.body.attestations.append(prev_attestation)
+
+        state_transition(post_state, block)
+        blocks.append(block)
+
+    return state, blocks, post_state
+
+
+def test_finality_from_genesis_rule_4(state):
```

Comment:
I understand what you mean by "from_genesis", but since there's no other rule 4 test case, it seems okay to move it to the last and rename it to `def test_finality_rule_4(state)`.